python/pyspark/sql/dataframe.py

```diff
@@ -1272,16 +1272,18 @@ def replace(self, to_replace, value, subset=None):
         """Returns a new :class:`DataFrame` replacing a value with another value.
         :func:`DataFrame.replace` and :func:`DataFrameNaFunctions.replace` are
         aliases of each other.
+        Values `to_replace` and `value` should be homogeneous. Mixed string and numeric
```

Comment:
We can leave the warning about truncation off if you think its unnecessary - but I do like the description you came up with for it.

I think the current proposed text on the expected types is a little vague (as you said) and I think we can have something that is precise, accurate, and easy to read so I'd like us to give it a shot :)

How about something like: "The element(s) `to_replace` and `value` should be the same type(s) (either all numerics, all booleans, or all strings)." This way we've clarified that mixing the different numerics is ok since it all gets converted to doubles at the end of the day? I'm open to other suggestions though too.